bothan-cryptocompare/src/worker/asset_worker.rs

```diff
@@ -1,132 +1,184 @@
+                error!("failed to add query ids to store")
             }
+        }
+    }
+}
+
+async fn handle_reconnect(
+    connector: &CryptoCompareWebSocketConnector,
+    connection: &mut CryptoCompareWebSocketConnection,
+    query_ids: &WorkerStore,
+) {
+    let mut retry_count: usize = 1;
+    loop {
+        warn!("reconnecting: attempt {}", retry_count);
+
+        if let Ok(new_connection) = connector.connect().await {
+            *connection = new_connection;
+
+            // Resubscribe to all ids
+            let Ok(ids) = query_ids.get_query_ids().await else {
+                error!("failed to get query ids from store");
+                return;
+            };
+
+            let ids_vec = ids.into_iter().collect::<Vec<String>>();
 
-            // Set multiple assets at once
-            if let Err(e) = store.set_assets(to_set).await {
-                warn!("failed to set multiple assets with error: {}", e);
+            if subscribe(&ids_vec, connection).await.is_ok() {
+                info!("resubscribed to all ids");
+            } else {
+                error!("failed to resubscribe to all ids");
             }
+        } else {
+            error!("failed to reconnect to binance");
         }
-        Err(e) => warn!("failed to get price data with error: {}", e),
+
+        retry_count += 1;
+        sleep(RECONNECT_BUFFER).await;
     }
 }
 
-fn parse_symbol_price(
-    id: &str,
-    symbol_price: &f64,
-    timestamp: &i64,
-) -> Result<AssetInfo, ParseError> {
-    let price_value = Decimal::from_f64(symbol_price.to_owned())
-        .ok_or(ParseError::InvalidPrice(symbol_price.to_owned()))?;
-    Ok(AssetInfo::new(
-        id.to_owned(),
-        price_value,
-        timestamp.to_owned(),
-    ))
+async fn store_ref_tick(store: &WorkerStore, ref_tick: ReferenceTick) -> Result<(), WorkerError> {
+    let id = ref_tick.instrument.clone();
+    let price =
+        Decimal::from_f64(ref_tick.value).ok_or(WorkerError::InvalidDecimal(ref_tick.value))?;
+
+    let asset_info = AssetInfo::new(ref_tick.instrument, price, ref_tick.last_update);
+    store.set_asset(id, asset_info).await?;
+
+    Ok(())
 }
 
-#[cfg(test)]
-mod test {
-    use super::*;
-    use rust_decimal::Decimal;
-    use std::str::FromStr;
-
-    #[test]
-    fn test_parse_symbol_price() {
-        let id = "BTC";
-        let price = 8426.69;
-        let timestamp = 1609459200;
-
-        // Call parse_symbol_price with the updated parameters
-        let result = parse_symbol_price(id, &price, &timestamp);
-        let expected = AssetInfo::new(
-            id.to_string(),
-            Decimal::from_str("8426.69").unwrap(),
-            timestamp,
-        );
-        assert_eq!(result.unwrap(), expected);
+async fn process_packet(store: &WorkerStore, packet: Packet) {
+    match packet {
+        Packet::RefTickAdaptive(ref_tick) => match store_ref_tick(store, ref_tick).await {
+            Ok(_) => info!("stored data"),
```

Comment:
[No change require]
just note that these log will be changed in the add-sources-api branch